mlir/lib/Dialect/Vector/Transforms/LowerVectorInterleave.cpp

```diff
@@ -0,0 +1,64 @@
+//===- LowerVectorInterleave.cpp - Lower 'vector.interleave' operation ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements target-independent rewrites and utilities to lower the
+// 'vector.interleave' operation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/Dialect/Vector/Transforms/LoweringPatterns.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+
+#define DEBUG_TYPE "vector-interleave-lowering"
+
+using namespace mlir;
+using namespace mlir::vector;
+
+namespace {
+/// Progressive lowering of InterleaveOp.
+class InterleaveOpLowering : public OpRewritePattern<vector::InterleaveOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::InterleaveOp op,
+                                PatternRewriter &rewriter) const override {
+    VectorType resultType = op.getResultVectorType();
+    // 1-D vector.interleave ops can be directly lowered to LLVM (later).
+    if (resultType.getRank() == 1)
+      return failure();
+
+    // Below we unroll the leading (or front) dimension. If that dimension is
+    // scalable we can't unroll it.
+    if (resultType.getScalableDims().front())
+      return failure();
+
+    // n-D case: Unroll the leading dimension.
+    auto loc = op.getLoc();
+    Value result = rewriter.create<arith::ConstantOp>(
+        loc, resultType, rewriter.getZeroAttr(resultType));
+    for (int idx = 0, end = resultType.getDimSize(0); idx < end; ++idx) {
+      Value extractLhs = rewriter.create<ExtractOp>(loc, op.getLhs(), idx);
+      Value extractRhs = rewriter.create<ExtractOp>(loc, op.getRhs(), idx);
+      Value interleave =
+          rewriter.create<InterleaveOp>(loc, extractLhs, extractRhs);
+      result = rewriter.create<InsertOp>(loc, interleave, result, idx);
+    }
+
+    rewriter.replaceOp(op, result);
```

Comment:
Could we make this rewrite iterative instead of recursive?

Reply:
It's naturally a recursive problem, so I feel like that'd complicate the rewrite (this is how all other lowerings like this hard handled too).

Reply:
Note that every time we unroll one dimension the op will go through all the canonicalization/folders the rewrite applies, which is expensive. Given that `ExtractOp` supports multiple indices, wouldn't it be possible to just go over all the dimensions we have to unroll? What am I missing? I'm actually surprised we don't have a utility to do something like that already.

Reply:
AFAIK there's no helper for unrolling vector types (correctly accounting for scalable dimensions). There's the stuff to build such a helper, so I've now added `vector::createUnrollIterator(vType, targetRank)` (to `VectorUtils.h`) and used that for this patch. 

But this is not the normal way these rewrites seem to work (for whatever reason). For example, the lowerings for both `vector.create_mask` and `vector.broadcast` just unroll a single dimension at a time.

Reply:
Sorry, I didn't intend to enforce this but the codebase is evolving quickly and I think this is one of those things we may want to challenge even if it was done differently somewhere else. Thanks for taking care of it.